### ISSUE TITLE
🙋 Deterministic builds

### ISSUE THREAD

**Issue by termhn** [parcel-bundler/parcel#735]
I want to use Parcel in my project [RippleWarpWallet](https://github.com/termhn/ripplewarpwallet) and possible future derivatives. One thing blocking that happening right now is that even given the same exact input code, building twice will produce slightly different outputs. It is really nice to be able to have a deterministic output given the same inputs because this lets people verify that the live online version of the tool is really the same as what is directly built from the source code (verifiable builds).

### 🎛 Configuration (.babelrc, package.json, cli command)

In my `package.json` I have this script which runs to do the preliminary parcel build

```json
    "parcelbuild": "parcel build index.html --no-minify --public-url ./ --out-dir ./build"
```

I then do some post processing in a script to directly insert the bundled code into the HTML document because this is convenient for my purposes.

### 🤔 Expected Behavior

Given the same input code, two builds one after the other should produce the same output code bundles.

### 😯 Current Behavior

There are small differences in the code that are insignificant to functionality but result in different SHA hashes and hits in diff tools, making verifying builds much harder for the average user. An example is these two builds which were done successively:

https://gist.github.com/termhn/ddaf437fc20fa50afe0378256869c901

Which have different SHA256 hashes and if compared in a text diffing tool like https://www.diffchecker.com/ result in several minor changes.

### 💁 Possible Solution

Not sure. Haven't yet dug into the code to see where these number changes are coming from or possible solutions.

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.5.1
| Node             | 8.1.3
| npm/Yarn         | yarn 0.27.5
| Operating System | Windows 10

**Comment by gamebox** [comment]
I'd also support looking into this as I just filed an issue(#753 ) for file "rev"ing which obviously requires the hash to remain consistent for the same input.

**Comment by termhn** [comment]
Right so I got a little chance to play around with the code. I figured out a way to get this to work, but it's sort of a patch on top... I haven't quite figured out the real source of the issue. The problem is that Asset IDs are not consistently assigned between builds, and the Asset ID is what is referenced for the require polyfills. This has something to do with the fact that when assets are not fully cached (and maybe even when they are for larger projects?), there are multiple duplicate assets created for the same files. For example, in a basic setup with an `index.html` that references `index.js` and then `index.js` references `include.js`, there are two distinct `Asset`s created for the `index.html`, three distinct assets created for `index.js` and two distinct assets created for `include.js`

Here's some output from adding logging statements inside the `Asset` constructor as well as in the `Bundler`'s `resolveAsset` method.
```
bundler resolving asset C:\Users\Gray\Code\test\index.html
new asset, name: C:\Users\Gray\Code\test\index.html id: 1
new asset, name: C:\Users\Gray\Code\test\index.html id: 2
new asset, name: C:\Users\Gray\Code\test\index.js id: 3
bundler resolving asset ./index.js
new asset, name: C:\Users\Gray\Code\test\index.js id: 4
new asset, name: C:\Users\Gray\Code\test\index.js id: 5
bundler resolving asset ./include.js
new asset, name: C:\Users\Gray\Code\test\include.js id: 6
new asset, name: C:\Users\Gray\Code\test\include.js id: 7
```
I originally suspected this had something to do with multithreading/workers and data races, but even with only one worker the same thing still persists. 

I solved the issue by changing `Asset.js` line 19 to
```
this.id = md5(name + pkg);
```
However this isn't ideal... first because it's not really solving the root issue here and second because for minified files you don't want to have a bunch of md5 hashes everywhere when it could just be a few numbers instead.
